git/libgit2/transport/ssh.go

```diff
@@ -0,0 +1,433 @@
+
+	port := "22"
+	if u.Port() != "" {
+		port = u.Port()
+	}
+	addr := net.JoinHostPort(u.Hostname(), port)
+
+	t.once.Do(func() {
+		if opts.Context != nil {
+			t.ctx = opts.Context
+			t.logger = ctrl.LoggerFrom(t.ctx,
+				"transportType", "ssh",
+				"addr", addr)
+		}
+	})
+
+	var cmd string
+	switch action {
+	case git2go.SmartServiceActionUploadpackLs, git2go.SmartServiceActionUploadpack:
+		if t.currentStream != nil {
+			if t.lastAction == git2go.SmartServiceActionUploadpackLs {
+				return t.currentStream, nil
+			}
+		}
+		cmd = fmt.Sprintf("git-upload-pack '%s'", uPath)
+
+	case git2go.SmartServiceActionReceivepackLs, git2go.SmartServiceActionReceivepack:
+		if t.currentStream != nil {
+			if t.lastAction == git2go.SmartServiceActionReceivepackLs {
+				return t.currentStream, nil
+			}
+		}
+		cmd = fmt.Sprintf("git-receive-pack '%s'", uPath)
+
+	default:
+		return nil, fmt.Errorf("unexpected action: %v", action)
+	}
+
+	sshConfig, err := createClientConfig(opts.AuthOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	sshConfig.HostKeyCallback = func(hostname string, _ net.Addr, key ssh.PublicKey) error {
+		keyHash := sha256.Sum256(key.Marshal())
+		return checkKnownHost(hostname, opts.AuthOpts.KnownHosts, keyHash[:])
+	}
+
+	if t.connected {
+		// The connection is no longer shared across actions, so ensures
+		// all has been released before starting a new connection.
+		_ = t.Close()
+	}
+
+	err = t.createConn(addr, sshConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	t.logger.V(logger.TraceLevel).Info("creating new ssh session")
+	if t.session, err = t.client.NewSession(); err != nil {
+		return nil, err
+	}
+
+	if t.stdin, err = t.session.StdinPipe(); err != nil {
+		return nil, err
+	}
+
+	var w *io.PipeWriter
+	var reader io.Reader
+	t.stdout, w = io.Pipe()
+	if reader, err = t.session.StdoutPipe(); err != nil {
+		return nil, err
+	}
+
+	// If the session's stdout pipe is not serviced fast
+	// enough it may cause the remote command to block.
+	//
+	// xref: https://github.com/golang/crypto/blob/eb4f295cb31f7fb5d52810411604a2638c9b19a2/ssh/session.go#L553-L558
+	go func() error {
+		defer func() {
+			w.Close()
+
+			// In case this goroutine panics, handle recovery.
+			if r := recover(); r != nil {
+				t.logger.V(logger.TraceLevel).Error(errors.New(r.(string)),
+					"recovered from libgit2 ssh smart subtransport panic")
+			}
+		}()
+		var cancel context.CancelFunc
+		ctx := t.ctx
+
+		// When context is nil, creates a new with internal SSH connection timeout.
+		if ctx == nil {
+			ctx, cancel = context.WithTimeout(context.Background(), sshConnectionTimeOut)
+			defer cancel()
+		}
+
+		closedAlready := atomic.LoadInt32(t.closedSessions)
+		for {
+			select {
+			case <-ctx.Done():
+				t.logger.V(logger.TraceLevel).Error(ctx.Err(), "context channel Done closed")
+				if atomic.LoadInt32(t.closedSessions) < closedAlready {
+					t.logger.V(logger.TraceLevel).Info("closing transport", "address", addr)
```

Comment:
I think `addr` is redundant here, as this is already set on line 175.
But adding `cmd` here could be useful when debugging cross-action or out-of-order closures.

```suggestion
					t.logger.V(logger.TraceLevel).Info("closing transport", "command", cmd)
```